iotfunctions/anomaly.py

```diff
@@ -507,15 +509,36 @@ def prepare_data(self, dfEntity):
     def execute(self, df):
 
         df_copy = df.copy()
-        entities = np.unique(df_copy.index.levels[0])
-        logger.debug(str(entities))
 
-        df_copy[self.output_item] = 0
-        # df_copy.sort_index() - NoOP
+        logger.info('Expression exp: ' + self.expression + '  input: ' + str(self.input_item))
+        expr = self.expression
+        if '${' in expr:
+            expr = re.sub(r"\$\{(\w+)\}", r"df['\1']", expr)
+            msg = 'Expression converted to %s. ' % expr
+        else:
+            msg = 'Expression (%s). ' % expr
+
+        self.trace_append(msg)
+
+        expr = str(expr)
+        logger.info('Expression - after regexp: ' + expr)
+
+        try:
+            mask = eval(expr)
+            entities = df_copy.loc[mask].index.unique(level=0)
+        except Exception as e:
+            logger.info('Expression eval for ' + expr + ' failed with ' + str(e))
+
+        df_copy[self.output_item] = np.nan
+
+        logger.debug('Entities to be processed {}'.format(str(entities)))
 
         for entity in entities:
             # per entity - copy for later inplace operations
-            dfe = df_copy.loc[[entity]].dropna(how='all')
+            entity_expr = "(" + "df['deviceid'] ==" + entity + ")"
+            entity_mask = eval(expr + '&' + entity_expr)
```

Comment:
I think it's not necessary to evaluate 'expr' again. You already have the result vector of 'expr' in 'mask'. Actually, the dynamic evaluation of "expr + '&' + entity_expr" is unnecessary overhead:
entity_expr = ( df['deviceid'] == entity)        (This returns a vector with True/False values)
entity_mask = (mask & entity_expr)               (This returns a vector with True values where mask  and entity_expr are true)